### PR TITLE
[SYCL][NFC] Detach compiler interfaces from PI

### PR DESCRIPTION
**Pull request by AlexeySachkov**
Our compiler interfaces do not depend on runtime bits and therefore they shouldn't be tied to PI, UR or anything else. This PR is a by-product of #14145 intended to refactor the code in advance so once we move to UR there are less changes required.

Moreover, compiler interfaces are not used anywhere in public SYCL headers and therefore they were moved into library.

As you can see, there are still some `pi` naming present in filenames and namespaces, but this patch is huge enough already, so further changes will be done as separate PRs.

### CODE REVIEW

File: sycl/include/sycl/detail/pi.h [hunk 1]
```diff
@@ -1369,6 +1173,11 @@ typedef struct {
 
 using pi_launch_property = _pi_launch_property;
 
+struct sycl_device_binaries_struct;
```
Comment (AlexeySachkov) [inline sycl/include/sycl/detail/pi.h:1176]: Strictly speaking, a couple of our compiler interfaces are part of the PI interface. But it won't be the case with UR if I understand correctly and therefore I went with those typedefs. There is also a few places (pi2ur, one unit-test and program manager) where reinterpret casts appeared to convert between `pi` alias and an actual type.

File: sycl/plugins/hip/CMakeLists.txt [hunk 2]
```diff
@@ -122,6 +122,7 @@ add_sycl_plugin(hip
     ${sycl_plugin_dir}
     ${CMAKE_CURRENT_SOURCE_DIR}/../unified_runtime
     ${UNIFIED_RUNTIME_SOURCE_DIR}/source/ # for adapters/hip
+    "${SYCL_SOURCE_DIR}/source" # for compiler.hpp
```
Comment (AlexeySachkov) [inline sycl/plugins/hip/CMakeLists.txt:125]: `pi2ur` depends on compiler interfaces and all plugins depend on `pi2ur`. Therefore, this is a necessary evil (?) for now.
My assumption is that this include won't be needed once we drop plugins in favor of UR.

File: sycl/plugins/cuda/CMakeLists.txt [hunk 3]
```diff
@@ -46,6 +46,8 @@ add_sycl_plugin(cuda
     ${XPTI_INCLUDE}
     ${CMAKE_CURRENT_SOURCE_DIR}/../unified_runtime # for Unified Runtime
     ${UNIFIED_RUNTIME_SOURCE_DIR}/source/ # for adapters/cuda
+    # Temporary (?) hack to include compiler interface header
```
Comment (AlexeySachkov) [inline sycl/plugins/cuda/CMakeLists.txt:49]:
```suggestion
```

File: sycl/source/detail/compiler.hpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
+//==--------- compiler.h - Interface between compiler and runtime ----------==//
```
Comment (AlexeySachkov) [inline sycl/source/detail/compiler.hpp:1]:
```suggestion
//==-------- compiler.hpp - Interface between compiler and runtime ---------==//
```

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 5]
```diff
@@ -535,8 +536,8 @@ static bool compatibleWithDevice(RTDeviceBinaryImage *BinImage,
   // compatible with implementation. The function returns invalid index if no
   // device images are compatible.
   pi_uint32 SuitableImageID = std::numeric_limits<pi_uint32>::max();
-  pi_device_binary DevBin =
-      const_cast<pi_device_binary>(&BinImage->getRawData());
+  auto DevBin = reinterpret_cast<pi_device_binary>(
```
Comment (AlexeySachkov) [inline sycl/source/detail/program_manager/program_manager.cpp:539]: An example of new `reinterpret_cast` I was talking in another comment.

File: sycl/unittests/SYCL2020/KernelBundleStateFiltering.cpp [hunk 6]
```diff
@@ -127,10 +127,11 @@ pi_result redefinedDevicesGet(pi_platform platform, pi_device_type device_type,
 }
 
 pi_result redefinedExtDeviceSelectBinary(pi_device device,
-                                         pi_device_binary *binaries,
+                                         pi_device_binary *pi_binaries,
                                          pi_uint32 num_binaries,
                                          pi_uint32 *selected_binary_ind) {
   EXPECT_EQ(num_binaries, 1U);
+  auto *binaries = reinterpret_cast<sycl_device_binary *>(pi_binaries);
```
Comment (AlexeySachkov) [inline sycl/unittests/SYCL2020/KernelBundleStateFiltering.cpp:134]: Another `reinterpret_cast` example

**Review by victor-eds — COMMENTED** [review]
KF changes LGTM

**Review by victor-eds — APPROVED** [review]

**Review by kbenzie — APPROVED** [review]

File: sycl/include/sycl/detail/pi.h (hunk 1 above)
Comment (kbenzie) [reply sycl/include/sycl/detail/pi.h:1176]: UR won't depend on these types.

File: sycl/plugins/cuda/CMakeLists.txt [hunk 7]
```diff
@@ -46,6 +46,8 @@ add_sycl_plugin(cuda
...
+    "${SYCL_SOURCE_DIR}/source"
```
Comment (AlexeySachkov) [inline sycl/plugins/cuda/CMakeLists.txt:50]:
```suggestion
    "${SYCL_SOURCE_DIR}/source" # for compiler.hpp
```

**Comment by AlexeySachkov** [comment]
@dpcpp-nativecpu-pi-reviewers, @jchlanda, could you please take a look? Changes in components you own are single-liners
@intel/llvm-reviewers-runtime, @cperkinsintel, @steffenlarsen, @aelovikov-intel, you are primary reviewers for this PR, could you please take a look? It is mostly a mechanical find-and-replace

**Review by jchlanda — APPROVED** [review]
CUDA/HIP 👍

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 8]
```diff
@@ -3136,12 +3141,12 @@ checkDevSupportDeviceRequirements(const device &Dev,
 } // namespace _V1
 } // namespace sycl
 
-extern "C" void __sycl_register_lib(pi_device_binaries desc) {
+extern "C" void __sycl_register_lib(sycl_device_binaries desc) {
```
Comment (cperkinsintel) [inline sycl/source/detail/program_manager/program_manager.cpp:3144]: I'm not necessarily asking for this PR, but can we document `__sycl_register_lib` somewhere?  It is one of the lynchpins between the FE and SYCL and has exactly zero comments or docs anywhere in the code base.

**Review by cperkinsintel — APPROVED** [review]

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 8 above)
Comment (AlexeySachkov) [reply sycl/source/detail/program_manager/program_manager.cpp:3144]: Yes, I agree that we a huge gap in our documentation, but I think that it should be a separate PR

**Comment by AlexeySachkov** [comment]
@PietroGhg, @uwedolinsky, could you please take a look? I'm missing review for a single line of CMakeLists.txt changes adding a path to more include headers

**Review by uwedolinsky — APPROVED** [review]
native_cpu looks ok

**Pull request merged**